src/beliefpropagation.jl

```diff
@@ -0,0 +1,244 @@
+#Construct the random initial Message Tensors for an ITensor Network, based on a partitioning into subgraphs specified ny 'sub graphs'
+#The ITensorNetwork needs to be flat (i.e. just sites and link indices, no site indices), and is assumed to only have 1 link indice between any two sites
+#id_init = 0 => Random, id_init = 1 => Identity Matrix initialisation (preferred)
+function construct_initial_mts(
+  g::NamedDimGraph, flatpsi::ITensorNetwork, dg_subgraphs::DataGraph; id_init=1
+)
+  mts = Dict{Pair,ITensor}()
+
+  for i in vertices(dg_subgraphs)
+    tns_to_contract = ITensor[]
+    for j in neighbors(dg_subgraphs, i)
+      edge_inds = []
+
+      for vertex in dg_subgraphs[i]
+        psiv = flatpsi[vertex]
+        for e in NamedDimEdge.(Ref(vertex) .=> neighbors(g, vertex))
+          if (find_subgraph(dst(e), dg_subgraphs) == j)
+            edge_ind = commoninds(flatpsi, e)[1]
+            push!(edge_inds, edge_ind)
+          end
+        end
+      end
+      if (id_init == 1)
+        A = Array(delta(edge_inds), edge_inds)
+        X1 = ITensor(A, edge_inds)
+      else
+        X1 = randomITensor(edge_inds)
+      end
+      normalize!(X1)
+      mts[i => j] = X1
+    end
+  end
+
+  return mts
+end
+
+#DO a single update of a message tensor using the current subgraph and the incoming mts
+function updatemt(
+  g::NamedDimGraph, flatpsi::ITensorNetwork, subgraph::Vector{Tuple}, mts::Vector{ITensor}
+)
+  Contract_list = ITensor[]
+
+  for m in mts
+    push!(Contract_list, m)
+  end
+
+  for v in subgraph
+    psiv = flatpsi[v]
+    push!(Contract_list, psiv)
+  end
+
+  M = ITensors.contract(Contract_list)
+
+  normalize!(M)
+
+  return M
+end
+
+#Do an update of all message tensors for a given flat ITensornetwork and its partition into sub graphs
+function update_all_mts(
+  g::NamedDimGraph,
+  flatpsi::ITensorNetwork,
+  mts::Dict{Pair,ITensor},
```

Comment:
A slight code simplification could be to store the message tensors as the edge metadata of the `dg_subgraphs` object. I think it makes sense to put them together in one object, since then all of the metadata/caching required for the generalized BP algorithm is in one place.

However, an issue I see is that the standard `DataGraph` is an undirected graph, and by default it enforces that the metadata for an edge in both directions is the same. That was a subtle design choice I went back and forth about, which makes sense for `IndsNetwork` but for this case I see it is too restrictive. It could be that we will need to use a `DataDiGraph` (though that doesn't exist right now I think).

Also, is there a reason to pass around `g`? Doesn't it contain the same graph information as `flatpsi`?

Reply:
Yeah that would be neat and makes a lot of sense to have dg as a date_di_graph. Would it be easy to define such a thing in DataGraphs.jl? 

Good point about g. All the relevant info should be in psi so I have removed the need for g in any of the belief propagation.jl

Reply:
It wouldn't be hard to define `DataDiGraph` but let's leave that for a future PR.